src/transformers/models/bark/modeling_bark.py

```diff
@@ -0,0 +1,1504 @@
+
+        Args:
+            input_ids (Optional[torch.Tensor] of shape (batch_size, seq_len), optional):
+                Input ids. Will be truncated up to 256 tokens. Note that the output audios will be as long as the
+                longest generation among the batch.
+            The last token is `semantic_infer_token`. Note that batch_size is set to 1 to generate one audio per audio. Defaults to None.:
+            history_prompt (Optional[Dict[str,torch.Tensor]], optional):
+                Optional `Bark` speaker prompt. Defaults to None. Note that for now, this model takes only one speaker
+                prompt per batch.
+            max_coarse_history (int, optional):
+                Max length of the output of the coarse acoustics model used in the fine generation step. Defaults to
+                630.
+            sliding_window_len (int, optional):
+                The coarse generation step uses a sliding window to generate raw audio. Defaults to 60.
+        Returns:
+            torch.LongTensor: Output generated audio.
+
+        Example:
+
+        ```python
+        >>> from transformers import AutoProcessor, BarkModel
+        >>> from scipy.io.wavfile import write as write_wav
+
+
+        >>> processor = AutoProcessor.from_pretrained("ylacombe/bark-small")
+        >>> model = BarkModel.from_pretrained("ylacombe/bark-small")
+
+        >>> inputs, _ = processor("Hello, my dog is cute, I need him in my life")
+
+        >>> audio_array = model.generate_audio(**inputs)
+        >>> audio_array = audio_array.detach().cpu().numpy()
+        ```
+
+        ```python
+        >>> # To add a voice preset, you can pass `voice_preset` to `BarkProcessor.__call__(...)`
+
+        >>> voice_preset = "v2/en_speaker_6"
+
+        >>> inputs, history_prompt = processor(
+        ...     "Hello, my dog is cute, I need him in my life", voice_preset=voice_preset
+        ... )
+
+        >>> audio_array = model.generate_audio(**inputs, history_prompt=history_prompt)
+        >>> audio_array = audio_array.detach().cpu().numpy().squeeze()
+
+        >>> # save audio to disk, but first take the sample rate from the model config
+        >>> sample_rate = model.config.sample_rate
+        >>> write_wav("bark_generation.wav", sample_rate, audio_array)
+        ```
+        """
+
+        ##### 1. Generate from the semantic model
+
+        semantic_output = self.generate_text_semantic(
+            input_ids, history_prompt, attention_mask=kwargs.pop("attention_mask", None), **kwargs
+        )
+
+        ##### 2. Generate from the coarse model
+
+        coarse_output = self.generate_coarse(semantic_output, history_prompt, **kwargs)
+
+        ##### 3. "generate" from the fine model
+
+        output = self.generate_fine(coarse_output, history_prompt, **kwargs)
+
+        #### 4. Decode the output and generate audio array
+
+        audio = self.codec_decode(output)
+
+        return audio
+
+
+# TODO: (maybe do it in the preprocessor)
```

Comment:
I would do this directly in the `preprocess_histories_before_coarse` method and remove this function